### PR TITLE
support to test sizes of some basic structures

### PR DESCRIPTION
**Pull request by mkeskells**

### CODE REVIEW

**Comment by retronym** [comment]
We java JOL available which lets us directly measure footprints. I'd rather use that unless you see a problem with it.

```
scala> val data = Array[String](null)
data: Array[String] = Array(null)

scala> import org.openjdk.jol.info.GraphLayout
import org.openjdk.jol.info.GraphLayout

scala> val netLayout = GraphLayout.parseInstance(data.toSeq).subtract(GraphLayout.parseInstance(data))
netLayout: org.openjdk.jol.info.GraphLayout = org.openjdk.jol.info.GraphLayout@627ad446

scala> netLayout.totalSize
res2: Long = 24

scala> println(netLayout.footPrint)
<console>:14: error: value footPrint is not a member of org.openjdk.jol.info.GraphLayout
       println(netLayout.footPrint)
                         ^

scala> println(netLayout.toFootprint)
 footprint:
     COUNT       AVG       SUM   DESCRIPTION
         1        24        24   scala.collection.mutable.WrappedArray$ofRef
         1                  24   (total)
```

**Comment by mkeskells** [comment]
> We java JOL available which lets us directly measure footprints. I'd rather use that unless you see a problem with it.

reworked to use JOL

File: test/junit/scala/collection/Sizes.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+object Sizes {
+  lazy val list:  Long = 24
+  lazy val listBuffer: Long = 32
+
+  def refArray(length:Int): Long = {
+    (16 + (length+1) * 4) /8 * 8
+  }
+  def wrappedRefArray(length:Int): Long = refArray(length) + 24
+
+}
```
Comment (retronym) [inline test/junit/scala/collection/Sizes.scala:18]: Where are you heading with this test? Is the intention to guard against accidental addition of `val`-s / `object`-s to core data structures?

That's okay, but I'd recommend using the `GraphLayout.subtract` idiom across the board to make an assertion about the net size of the data structure.

```scala
def netLayout(wrappper: Object, wrapped: Object): GraphLayout = GraphLayout.parseInstance(wrapper).substract(GraphLayout.parseInstance(wrapped)

def assertNetSize(expectedSize: Int, wrappper: Object, wrapped: Object): Unit = {
   val layout = netLayout(wrapper, wrapped)
   assertEquals(layout.toPrintable, expectedSize, layout.totalSize)
}
...

assertNetSize(24, someArray.toSeq, someArray))
```

**Comment by retronym** [comment]
I'll merge and rework this slightly as part of #8842

**Pull request closed without merging**